File: trunk/plsi_camp/reduce.f90
program reduce
include 'mpif.h'

integer ierr, myrank, nprocs, ista, iend, a(10000)
real sum, tmp

call MPI_INIT(ierr)
call MPI_COMM_RANK(MPI_COMM_WORLD, myrank, ierr)
call MPI_COMM_SIZE(MPI_COMM_WORLD, nprocs, ierr)

ista = myrank * 5000 + 1
iend = ista + 4999

do i = ista, iend
a(i) = i
enddo

sum = 0.0

do i = ista, iend
sum = sum + a(i)
enddo

call MPI_REDUCE(sum, tmp, 1, MPI_REAL, MPI_SUM, 0, MPI_COMM_WORLD, ierr)

sum = tmp

if (myrank == 0) then
    print *, 'sum =', sum
endif

!print *, 'rank =', myrank, ', a =', a

call MPI_FINALIZE(ierr)

end

File: trunk/plsi_camp/reduce.c
#include <stdio.h>
#include <mpi.h>

int main(int argc, char *argv[])
{
    int myrank, i, ista, iend;
    double a[9], sum, tmp;

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &myrank);

    ista = myrank * 3;
    iend = ista + 2;

    for (i = ista; i < iend + 1; i++) a[i] = i;

    sum = 0.0;

    MPI_Reduce(&sum, &tmp, 1, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD);

    sum = tmp;

    if (myrank == 0)
    {
        printf("sum = %f\n", sum);
    }

    MPI_Finalize();

    return 0;
}
